src/ssl/cert_validate_message.cc

```diff
@@ -144,8 +149,34 @@ Ssl::CertValidationMsg::tryParsingResponse(CertValidationResponse &resp)
         size_t value_len = strcspn(value, "\r\n");
         std::string v(value, value_len);
 
-        debugs(83, 5, "Returned value: " << std::string(param, param_len).c_str() << ": " <<
-               v.c_str());
+        debugs(83, 5, "Returned value: " << SBuf(param, param_len) << ": " << v);
+
+        if (param_len == param_transactionNotes.length() &&
+            strncmp(param, param_transactionNotes.c_str(), param_transactionNotes.length()) == 0) {
+            resp.notes.importFromHelper(SBuf(v.c_str(), v.length()));
+            param = value + value_len;
+            continue;
+        }
+
+        if (param_len == param_clientNotes.length() &&
+            strncmp(param, param_clientNotes.c_str(), param_clientNotes.length()) == 0) {
+
+            // TODO: Support arbitrary client annotations when up-porting.
+            static std::string supportedName = "clt_conn_tag=";
+            if (v.compare(0, supportedName.size(), supportedName) != 0) {
+                throw TextException(ToSBuf("Only annotations named ", supportedName,
+                    " can be used for client connection annotation in this Squid version. ",
+                    "Found: ", v), Here());
+            }
+            if (v.find(' ') != std::string::npos) {
+                throw TextException(ToSBuf("Only one client connection annotation can be used in this Squid version. ",
+                    "Found: ", v), Here());
+            }
+            resp.notes.importFromHelper(SBuf(v.c_str(), v.length()));
```

Comment:
This general-purpose-looking method looks misplaced - I guess it should be called earlier, e.g., from (or instead of) Helper::Reply::parseResponseKeys(), which it essentially duplicates. There are some differences, however, making this refactoring problematic:
* parseResponseKeys() expects fields that are not kv-pairs (and simply returns in this case), whereas importFromHelper() threats them as errors and throws.
* parseResponseKeys() expects quotes in the value part and does unescaping, while importFromHelper() does not do that.

Reply:
I would wait for this refactoring until later in this PR lifecycle, when internal importFromHelper() changes are completed, but here are a few notes for that future work (based on your comment alone -- I have not checked the actual code):

> * parseResponseKeys() expects fields that are not kv-pairs (and simply returns in this case), whereas importFromHelper() threats them as errors and throws.

This difference seems easy to reconcile: The combined method should successfully return on the first non-key=value field (or end of input). Callers that do not expect anything after key=value fields should call a second/new method to check for any trailing garbage.

> * parseResponseKeys() expects quotes in the value part and does unescaping, while importFromHelper() does not do that.

Ideally, we should support quotes, applying existing unescaping functionality to the certificate validator annotations. The problem here may stem from the fact that validator annotations are grouped inside a "quoted value" already. I do not think we should support nested quotes of the same type. I suggest using a different quoting mechanism for grouping validator annotations. I can suggest two options:

1. length-value (already used for the entire validator message): `annotations[n]=...exactly n bytes...`
2. multi-character quotations marks (already used for certificates): `annotations=-----BEGIN NOTES-----...-----END NOTES-----`

Despite it being less "human-friendly", I prefer the length-value approach because it is simpler, more robust, and less noisy.